### DISCUSSION TITLE
Data type of module params

### DISCUSSION THREAD

**Discussion by matthias-wright** [google/flax#1346] [Q&A] [answered]
Hi all,

I have noticed that the data type of common module params (linen.Conv, linen.Dense, etc) are initialized as float32 regardless of the `dtype` argument that is passed when initializing the module. 

```python
import jax
import jax.numpy as jnp
import flax.linen as nn

rng = jax.random.PRNGKey(0)

dtype = jnp.float16
model = nn.Conv(features=32, kernel_size=(3, 3), dtype=dtype)
params = model.init(rng, jnp.ones((1, 28, 28, 3), dtype=dtype))

print(params['params']['kernel'].dtype) # float32
```

Then in the `__call__` method, after the params are retrieved with `self.param(...)`, they are casted to `dtype` before computing the modules operation.

In case of  `dtype=float16`, wouldn't we save memory by initializing the params as `float16` directly and then "storing" them in the module as `float16`?

Thanks!

**Comment by jheek** [comment, marked as answer]
You can manually cast the weights to a half float after initialization like this:
```
params_f16 = jax.tree_map(lambda p: jnp.asarray(p, jnp.float16), params)
```

This is a cheap win for inference but for training it is not so obvious. Half floats suffer from major underflow (mainly float16) and addition errors (mainly bfloat16).  Which makes it hard to optimize them directly. Typically you get most of the benefits of half types by only storing activations in half precision but keeping the weights in float32 to stabilize training.

**Reply by pharringtonp19** [reply]
Is there a flax example of this?
> storing activations in half precision but keeping the weights in float32

**Reply by matthias-wright** [reply]
Check out [this](https://github.com/google/flax/blob/main/examples/imagenet/train.py) example. There is a `half_precision` flag that you can set.

**Reply by moabarar** [reply]
> You can manually cast the weights to a half float after initialization like this:
> 
> ```
> params_f16 = jax.tree_map(lambda p: jnp.asarray(p, jnp.float16), params)
> ```
> 
> This is a cheap win for inference but for training it is not so obvious. Half floats suffer from major underflow (mainly float16) and addition errors (mainly bfloat16). Which makes it hard to optimize them directly. Typically you get most of the benefits of half types by only storing activations in half precision but keeping the weights in float32 to stabilize training.

 I am currently implementing a new layer with Flax.linen.Module, and I follow the current internal implementation of flax, where immediately after the initialization of the module params I wrap the param using jnp.asarray(data, self.dtype)
For example:

```
kernel = self.param('Kernel', init, shape)
kernel = jnp.asarray(kernel, self.dype)
```

Would you say this is best practice ?
